scripts/projects-test/src/performBrowserTest.ts

```diff
@@ -1,54 +1,92 @@
-import http from 'http';
+import { Server } from 'http';
 import { AddressInfo } from 'net';
 
-import { safeLaunchOptions } from '@fluentui/scripts-puppeteer';
-import express from 'express';
-import puppeteer from 'puppeteer';
+import { launch, visitUrl } from '@fluentui/scripts-puppeteer';
+import express, { Express } from 'express';
 
+/**
+ *
+ * .close is asynchronous (not promisified thus we are wrapping it)
+ *  @see https://nodejs.org/api/net.html#serverclosecallback
+ */
+export function closeServer(server: Server): Promise<void> {
+  if (!server) {
+    return Promise.resolve();
+  }
 
-const SERVER_HOST = 'localhost';
+  return new Promise<void>((resolve, reject) => {
+    // TODO: Default is 5 seconds. why is this set to 1 ?
+    // https://nodejs.org/api/http.html#serverkeepalivetimeout
+    server.keepAliveTimeout = 1000;
+    server.close(err => (err ? reject(err) : resolve()));
+  });
+}
 
-function startServer(publicDirectory: string, listenPort: number) {
-  return new Promise<http.Server>((resolve, reject) => {
+export function startServer(
+  options: { root: string; port: number; host?: string },
+  configureMiddleware = (app: Express) => app,
+) {
+  const { root, port, host = 'localhost' } = options;
+  return new Promise<{ server: Server; port: number; url: string }>((resolve, reject) => {
+    let usedPort: number;
     try {
-      const app = express();
-      app.use(express.static(publicDirectory));
+      console.log('express: starting server');
+
+      const middleware = (app: Express) => app.use(express.static(root));
+      const app = middleware(configureMiddleware(express()));
+      const server = app.listen(port, host, () => {
+        const shouldAssignArbitraryUnusedPort = port === 0;
+        usedPort = shouldAssignArbitraryUnusedPort ? (server.address() as AddressInfo).port : port;
+        const url = `http://${host}:${usedPort}`;
+
+        console.log(`express: server running at http://${host}:${usedPort} from directory "${root}"`);
 
-      const server = app.listen(listenPort, SERVER_HOST, () => {
-        resolve(server);
+        resolve({ server, port: usedPort, url });
+      });
+
+      server.on('error', err => {
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore - improper Error type in typings -> https://nodejs.org/api/net.html#serverlisten
+        if (err.code === 'EADDRINUSE') {
```

Comment:
should we be a bit smarter handling this error when the port is randomly assigned by doing a retry with another randomly assigned port?